codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java

```diff
@@ -256,18 +270,24 @@ public final void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
      */
     protected void handlerRemoved0(ChannelHandlerContext ctx) throws Exception { }
 
+    /**
+     * Tells whether this handle has been removed from the {@link ChannelPipeline}.
+     * This method is useful because 1) a user can check if this handler is removed even if
+     * he or she does not have a reference to {@link ChannelHandlerContext} and 2) it is
+     * cheaper than {@link ChannelHandlerContext#isRemoved()}.
+     */
+    protected final boolean isRemoved() {
+        return decodeState >= STATE_HANDLER_REMOVED_PENDING;
+    }
+
     @Override
     public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
         if (msg instanceof ByteBuf) {
             CodecOutputList out = CodecOutputList.newInstance();
             try {
-                ByteBuf data = (ByteBuf) msg;
                 first = cumulation == null;
-                if (first) {
-                    cumulation = data;
-                } else {
-                    cumulation = cumulator.cumulate(ctx.alloc(), cumulation, data);
-                }
+                cumulation = cumulator.cumulate(ctx.alloc(),
+                        first ? Unpooled.EMPTY_BUFFER : cumulation, (ByteBuf) msg);
```

Comment:
Although `Unpooled.EMPTY_BUFFER` does not have any additional overhead, I still feel that this way of use is not easy to understand. Why use `Unpooled.EMPTY_BUFFER`? This actually depends on the logic of the `cumulate` method, which is a bit more complicated than the previous use.

So, personally prefer the original way. WDYT? :)

Reply:
if use the Unpooled.EMPTY_BUFFER then it should be ignored in the accumulator

Reply:
@carryxyh @hepin1989 the purpose of this is to give the cumulator a chance to replace the input buffer if appropriate before it's passed directly to the decoder. The primary motivation is so that we don't pass a composite buffer when `MERGE_CUMULATOR` is configured, as would currently be the case. Doing it this way ensures compatibility with any custom `Cumulator` impls that might be out there (though I would be surprised if any do really exist).